### PR TITLE
test(checkpoint): redesign resume robustness around a shared trajectory

### PR DESCRIPTION
**Pull request by yuhezhang-ai**
# What does this PR do?

Redesign checkpoint-resume robustness so resumed training is compared with the uninterrupted continuation of the same checkpoint-producing trajectory. Independent-run reproducibility remains visible as a separately named, non-blocking, opportunistic diagnostic that reuses existing training phases instead of launching another baseline run.

The previous resume oracle compared independently initialized runs, which conflated checkpoint restoration correctness with broader training reproducibility. This made resume thresholds recipe-specific and could hide missing optimizer, scheduler, RNG, or dataloader restoration behind a loose loss tolerance.

# Changelog

- Continue checkpoint Phase 1 past the checkpoint boundary and compare a fresh restore against the same per-rank post-boundary batches and losses.
- Validate exact optimizer step/group state, LR scheduler state, RNG state, LR, and weight decay before continuing training; verify stateful-dataloader position behaviorally through exact per-rank resumed batch identity.
- Use a strict first-post-resume loss threshold (`1e-6` by default) and a separate later-step BF16 envelope (`5e-3` by default). The distributed retrieval contrastive-loss recipe uses a calibrated `5e-2` later-step envelope while keeping its first resumed step strict.
- Remove the independently trained resume-baseline phase from CI generation and process-isolated execution.
- Record the existing normal finetune and checkpoint Phase 1 for independent-run reproducibility without an additional training run.
- Compare independent runs only when model/init, seed, data ordering, batch/topology, optimizer, scheduler, loss, and backend fingerprints match; phase-specific differences report `not_comparable` rather than pretending to provide coverage.
- Retain the previously calibrated recipe thresholds under `training_reproducibility_loss_threshold`; comparable-run violations emit a prominent non-blocking alert and a machine-readable `report.json`.
- Calibrate Qwen source-load parity and retrieval later-step resume envelopes from scoped CI observations without changing rank-local RNG persistence in this PR.
- Cover the shared trajectory, omitted restored state, strict first-step tolerance, configuration mismatch, batch divergence, loss-envelope alerting, CI config resolution, and phase generation with focused unit tests.
- Update checkpoint-robustness and CI documentation.

# Validation

Local validation:

- `pytest tests/unit_tests/ci_tests/test_resume_trajectory.py tests/unit_tests/ci_tests/test_config_resolver.py tests/unit_tests/ci_tests/test_generate_ci_tests.py -q` — 61 passed, 2 warnings
- Final-head `tests/unit_tests/ci_tests/test_config_resolver.py` suite — 41 passed
- Ruff and `git diff --check` — passed

Scoped NeMo CI:

- [Qwen VLM pipeline 61321539](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/61321539): [`qwen3_vl_moe_30b_te_deepep`](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/386704786) passed.
- [Retrieval pipeline 61327708](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/61327708): [`nemotron_vl_1b_example`](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/386767100) passed on current head `0765fc0b8cac898b294075d28bc8550c64e39172`.
- [Combined shared-trajectory + ranked-RNG pipeline 61326392](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/61326392): [`llama3_3_nemotron_super_49B_squad_peft`](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/386746385) passed the exact restored-state and shared-trajectory checks.

The scoped launches were restricted to EOS/x86, `BUILD_MODULE_REGEX=/^automodel$/`, and the named recipes. The retrieval generator's unused cross-encoder bridge had no selected test; the requested bi-encoder job passed.

# Pre checks

- [x] Read and followed the contributor guidelines
- [x] Added focused tests
- [x] Updated the relevant documentation
- [x] Confirmed the final scoped recipe jobs pass

# Additional Information

- Linear: [AMINT-227](https://linear.app/nvidia/issue/AMINT-227/redesign-checkpoint-resume-robustness-around-a-shared-training)
- Rank-local RNG checkpoint persistence is intentionally fixed separately in [PR #3437](https://github.com/NVIDIA-NeMo/Automodel/pull/3437); the combined PEFT run above validates the two changes together.

### CODE REVIEW

**Comment by yuhezhang-ai** [comment]
/ok to test [4e05ca1](https://github.com/NVIDIA-NeMo/Automodel/pull/3427/commits/4e05ca1f3e77f98e824c58ba53968292e82c6b01)

**Comment by yuhezhang-ai** [comment]
/ok to test [0765fc0](https://github.com/NVIDIA-NeMo/Automodel/pull/3427/commits/0765fc0b8cac898b294075d28bc8550c64e39172)

**Review by athitten — APPROVED** [review]
Thank you @yuhezhang-ai ! LGTM

**Review by thomasdhc — APPROVED** [review]
LGTM within the requested scope of CI test generation and the overall CI/CD flow.

### Review notes
- CI generation and launcher orchestration consistently remove the independent `resume_baseline` phase while preserving the shared-trajectory resume path.
- The normal finetune and checkpoint flow records and propagates reproducibility artifacts without changing blocking failure behavior.
